File: .skills/openclaw-skills/skills/0xnyk/xint/lib/article.ts
/**
 * lib/article.ts — Full article content fetcher via xAI Responses API.
 *
 * Uses Grok's web_search tool to fetch and extract article content
 * from URLs. No scraping — uses the xAI API which can access pages
 * that plain HTTP can't (JS-rendered, some paywalled content).
 *
 * Requires XAI_API_KEY in env or .env.
 */

import { readFileSync } from "fs";
import { join } from "path";
import { extractTweetId } from "./media";
import type { TweetArticle } from "./api";
import { trackCostDirect } from "./costs";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface Article {
  url: string;
  title: string;
  description: string;
  content: string;       // extracted article text
  author: string;
  published: string;     // date string
  domain: string;
  ttr: number;           // time to read in minutes
  wordCount: number;
}

interface ResponsesApiResult {
  id: string;
  output: Array<{
    type: string;
    content?: Array<{ type: string; text?: string }>;
    text?: string;
  }>;
  usage?: {
    input_tokens: number;
    output_tokens: number;
    total_tokens: number;
  };
}

// ---------------------------------------------------------------------------
// Auth
// ---------------------------------------------------------------------------

function getXaiKey(): string {
  if (process.env.XAI_API_KEY) return process.env.XAI_API_KEY;

  try {
    const envFile = readFileSync(join(import.meta.dir, "..", ".env"), "utf-8");
    const match = envFile.match(/XAI_API_KEY=["']?([^"'\n]+)/);
    if (match) return match[1];
  } catch {}

  throw new Error(
    "XAI_API_KEY not found. Set it in your environment or in .env"
  );
}

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const XAI_RESPONSES_ENDPOINT = "https://api.x.ai/v1/responses";
const DEFAULT_MODEL = "grok-4";
const DEFAULT_TIMEOUT_SEC = 30;

const ARTICLE_EXTRACT_PROMPT = `Read the article at this URL and extract its content. Return a JSON object with these fields:
- title: article title
- description: 1-2 sentence summary
- content: the full article text (plain text, no HTML)
- author: author name (empty string if unknown)
- published: publication date (empty string if unknown)

Return ONLY valid JSON, no markdown fences, no explanation.`;

// ---------------------------------------------------------------------------
// Fetcher
// ---------------------------------------------------------------------------

/**
 * Fetch and extract article content from a URL using xAI's web_search tool.
 */
export async function fetchArticle(
  url: string,
  opts: { model?: string; full?: boolean } = {},
): Promise<Article> {
  let parsed: URL;
  try {
    parsed = new URL(url);
  } catch {
    throw new Error(`Invalid URL: ${url}`);
  }

  const apiKey = getXaiKey();
  const model = opts.model || DEFAULT_MODEL;
  const timeoutMs = resolveArticleTimeoutMs();
  const timeoutSec = Math.floor(timeoutMs / 1000);
  const controller = new AbortController();
  const timeoutId = setTimeout(() => controller.abort("article_timeout"), timeoutMs);

  let res: Response;
  try {
    res = await fetch(XAI_RESPONSES_ENDPOINT, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${apiKey}`,
      },
      signal: controller.signal,
      body: JSON.stringify({
        model,
        tools: [
          {
            type: "web_search",
            allowed_domains: [parsed.hostname],
          },
        ],
        input: [
          {
            role: "user",
            content: `${ARTICLE_EXTRACT_PROMPT}\n\nURL: ${url}`,
          },
        ],
      }),
    });
  } catch (error: unknown) {
    if (controller.signal.aborted) {
      throw new Error(
        `Article fetch timed out after ${timeoutSec}s for ${url}. ` +
        `Set XINT_ARTICLE_TIMEOUT_SEC (5-120) to tune this.`
      );
    }
    throw new Error(`Article fetch request failed for ${url}: ${(error as Error).message}`);
  } finally {
    clearTimeout(timeoutId);
  }

  if (!res.ok) {
    const body = await res.text();
    throw new Error(`xAI API error (${res.status}): ${body.slice(0, 300)}`);
  }

  const data = (await res.json()) as ResponsesApiResult;

  // Track xAI Responses API cost
  if (data.usage) {
    const inputCost = (data.usage.input_tokens / 1_000_000) * 3.0;   // grok-4 input rate
    const outputCost = (data.usage.output_tokens / 1_000_000) * 15.0; // grok-4 output rate
    trackCostDirect("xai_article", XAI_RESPONSES_ENDPOINT, inputCost + outputCost);
  }

  // Extract text from response output
  const text = extractResponseText(data);
  if (!text) {
    throw new Error(
      `No article content returned for ${url}. ` +
      `The source may be blocked/unavailable from this environment.`
    );
  }

  // Parse the JSON response from Grok
  const article = parseArticleJson(text, url, parsed.hostname);

  // Truncate content unless --full
  if (!opts.full && article.content.length > 5000) {
    article.content = article.content.slice(0, 5000).replace(/\s+\S*$/, "") + "\n\n[... truncated]";
  }

  // Compute word stats from full content (before truncation for accurate count)
  const words = article.content.replace(/\[... truncated\]$/, "").split(/\s+/).filter(Boolean).length;
  article.wordCount = words;
  article.ttr = Math.ceil(words / 238);

  return article;
}

export function resolveArticleTimeoutMs(): number {
  const raw = Number.parseInt(String(process.env.XINT_ARTICLE_TIMEOUT_SEC || DEFAULT_TIMEOUT_SEC), 10);
  if (!Number.isFinite(raw)) return DEFAULT_TIMEOUT_SEC * 1000;
  const clamped = Math.min(Math.max(raw, 5), 120);
  return clamped * 1000;
}

// ---------------------------------------------------------------------------
// Response parsing
// ---------------------------------------------------------------------------

function extractResponseText(data: ResponsesApiResult): string | null {
  for (const output of data.output || []) {
    // Responses API: message type with content array
    if (output.type === "message" && output.content) {
      for (const block of output.content) {
        if (block.type === "output_text" && block.text) return block.text;
        if (block.type === "text" && block.text) return block.text;
      }
    }
    // Direct text field
    if (output.text) return output.text;
  }
  return null;
}

function parseArticleJson(raw: string, url: string, domain: string): Article {
  // Strip markdown fences if present
  let cleaned = raw.trim();
  if (cleaned.startsWith("```")) {
    cleaned = cleaned.replace(/^```(?:json)?\n?/, "").replace(/\n?```$/, "");
  }

  try {
    const parsed = JSON.parse(cleaned);
    return {
      url,
      title: parsed.title || domain,
      description: parsed.description || "",
      content: parsed.content || "",
      author: parsed.author || "",
      published: parsed.published || "",
      domain,
      ttr: 0,
      wordCount: 0,
    };
  } catch {
    // If Grok returned plain text instead of JSON, use it as content
    return {
      url,
      title: domain,
      description: "",
      content: cleaned,
      author: "",
      published: "",
      domain,
      ttr: 0,
      wordCount: 0,
    };
  }
}

// ---------------------------------------------------------------------------
// X Tweet URL to Article
// ---------------------------------------------------------------------------

/**
 * Extract tweet ID from X URL and fetch the tweet to get linked articles.
 */
export async function fetchTweetForArticle(tweetUrl: string): Promise<{
  tweet: any;
  articleUrl: string | null;
  inlineArticle?: Article;
}> {
  const tweetId = extractTweetId(tweetUrl);
  if (!tweetId) {
    throw new Error(`Invalid X tweet URL: ${tweetUrl}`);
  }

  // Import dynamically to avoid circular dependencies
  const { getTweet } = await import("./api");
  const tweet = await getTweet(tweetId);

  if (!tweet) {
    throw new Error(`Tweet not found: ${tweetId}`);
  }

  // If tweet has inline article data from X API, build Article directly
  if (tweet.article?.plain_text) {
    const content = reconstructArticleContent(tweet.article);
    const words = content.split(/\s+/).filter(Boolean).length;
    return {
      tweet,
      articleUrl: null,
      inlineArticle: {
        url: tweetUrl,
        title: tweet.article.title || "X Article",
        description: tweet.article.preview_text || "",
        content,
        author: tweet.username || tweet.name || "",
        published: tweet.created_at || "",
        domain: "x.com",
        ttr: Math.max(1, Math.ceil(words / 238)),
        wordCount: words,
      },
    };
  }

  return { tweet, articleUrl: pickArticleUrlFromTweet(tweet) };
}

export function reconstructArticleContent(article: TweetArticle): string {
  let content = article.plain_text;
  const codeBlocks = article.entities?.code;
  if (codeBlocks?.length) {
    content += "\n\n---\n\nCode examples from article:\n";
    for (const block of codeBlocks) {
      content += `\n${block.content}\n`;
    }
  }
  return content;
}

export function buildArticleFallbackFromTweet(input: {
  sourceUrl: string;
  tweet: any;
  articleTitle?: string;
  articleDescription?: string;
  reason: string;
}): Article {
  const title = input.articleTitle || "X Article Link";
  const description = input.articleDescription || input.reason;
  const tweetText = String(input.tweet?.text || "").trim();
  const content = [
    "Unable to fetch full article body in this runtime.",
    input.reason,
    tweetText ? `Tweet text: ${tweetText}` : "",
    `Source URL: ${input.sourceUrl}`,
  ].filter(Boolean).join("\n\n");

  const words = content.split(/\s+/).filter(Boolean).length;
  const domain = (() => {
    try {
      return new URL(input.sourceUrl).hostname;
    } catch {
      return "x.com";
    }
  })();

  return {
    url: input.sourceUrl,
    title,
    description,
    content,
    author: "",
    published: "",
    domain,
    ttr: Math.max(1, Math.ceil(words / 238)),
    wordCount: words,
  };
}

function normalizeCandidateUrl(value: string | undefined): string | null {
  if (!value || typeof value !== "string") return null;
  try {
    const parsed = new URL(value);
    if (parsed.protocol !== "http:" && parsed.protocol !== "https:") return null;
    return parsed.toString();
  } catch {
    return null;
  }
}

export function isXArticleUrl(url: string): boolean {
  try {
    const parsed = new URL(url);
    return /(^|\.)x\.com$/i.test(parsed.hostname) && parsed.pathname.startsWith("/i/article/");
  } catch {
    return false;
  }
}

function isExternalNonXUrl(url: string): boolean {
  try {
    const parsed = new URL(url);
    return !/(^|\.)x\.com$/i.test(parsed.hostname) && !/(^|\.)twitter\.com$/i.test(parsed.hostname);
  } catch {
    return false;
  }
}

/**
 * Pick the best article URL candidate from parsed tweet payload.
 * Priority:
 * 1) external non-X URLs
 * 2) X Article URLs (x.com/i/article/*)
 * 3) any remaining valid URL candidate
 */
export function pickArticleUrlFromTweet(tweet: any): string | null {
  const candidates: string[] = [];

  const parsedUrls = Array.isArray(tweet?.urls) ? tweet.urls : [];
  for (const urlData of parsedUrls) {
    const fromUnwound = normalizeCandidateUrl(urlData?.unwound_url);
    const fromUrl = normalizeCandidateUrl(urlData?.url);
    if (fromUnwound) candidates.push(fromUnwound);
    if (fromUrl) candidates.push(fromUrl);
  }

  const entityUrls = Array.isArray(tweet?.entities?.urls) ? tweet.entities.urls : [];
  for (const urlData of entityUrls) {
    const fromUnwound = normalizeCandidateUrl(urlData?.unwound_url);
    const fromExpanded = normalizeCandidateUrl(urlData?.expanded_url);
    const fromUrl = normalizeCandidateUrl(urlData?.url);
    if (fromUnwound) candidates.push(fromUnwound);
    if (fromExpanded) candidates.push(fromExpanded);
    if (fromUrl) candidates.push(fromUrl);
  }

  const seen = new Set<string>();
  const unique = candidates.filter((candidate) => {
    if (seen.has(candidate)) return false;
    seen.add(candidate);
    return true;
  });

  const external = unique.find(isExternalNonXUrl);
  if (external) return external;

  const xArticle = unique.find(isXArticleUrl);
  if (xArticle) return xArticle;

  return unique[0] || null;
}

// ---------------------------------------------------------------------------
// Formatting
// ---------------------------------------------------------------------------

/**
 * Format article for terminal display.
 */
export function formatArticle(article: Article): string {
  let out = `📰 ${article.title}\n`;
  if (article.author) out += `✍️  ${article.author}`;
  if (article.published) {
    const date = article.published.includes("T")
      ? article.published.split("T")[0]
      : article.published;
    out += out.endsWith("\n") ? "" : " · ";
    out += `${date}`;
  }
  if (article.author || article.published) out += "\n";
  out += `🔗 ${article.url}\n`;
  out += `📊 ${article.wordCount} words · ${article.ttr} min read\n`;
  if (article.description) {
    out += `\n${article.description}\n`;
  }
  out += `\n---\n\n${article.content}`;
  return out;
}

/**
 * Format article with tweet context for display.
 */
export function formatArticleWithTweet(article: Article, tweet: any): string {
  const tweetUrl = tweet.tweet_url || `https://x.com/${tweet.username}/status/${tweet.id}`;
  
  let out = `📝 Tweet: ${tweetUrl}\n`;
  out += `   ${tweet.text?.slice(0, 200)}${tweet.text?.length > 200 ? "..." : ""}\n\n`;
  out += formatArticle(article);
  
  return out;
}

File: .skills/openclaw-skills/skills/0xnyk/xint-rs/src/commands/article.rs
use anyhow::Result;

use crate::api::grok;
use crate::api::twitter;
use crate::api::xai;
use crate::cli::ArticleArgs;
use crate::client::XClient;
use crate::config::Config;
use crate::format;
use crate::models::{Article, Tweet, TweetArticle};

pub async fn run(args: &ArticleArgs, config: &Config) -> Result<()> {
    let mut url = args.url.clone();
    let mut article: Option<Article> = None;

    if is_x_tweet_like_url(&url) {
        eprintln!("Fetching tweet to extract linked article...");
        let token = config.require_bearer_token()?;
        let client = XClient::new()?;
        let (tweet, article_url, inline_article) =
            fetch_tweet_for_article(&client, token, &url).await?;

        if let Some(inline) = inline_article {
            eprintln!("Found X Article: {}\n", inline.title);
            article = Some(inline);
        } else if let Some(found) = article_url {
            eprintln!("Tweet: {}", tweet.tweet_url);
            eprintln!("Found link: {found}\n");
            url = found;
        } else {
            eprintln!("No external link found in tweet.");
            eprintln!("   Tweet: {}", tweet.tweet_url);
            return Ok(());
        }
    }

    let article = if let Some(a) = article {
        a
    } else if is_x_article_url(&url) {
        // X Articles can't be fetched via web_search (Grok can't browse x.com).
        // Build a metadata-only article and optionally enrich via Grok analyze.
        let meta_article = Article {
            url: url.clone(),
            title: "X Article".to_string(),
            description: String::new(),
            content: "[Article content is behind X authentication. Only metadata is available from the API.]".to_string(),
            author: String::new(),
            published: String::new(),
            domain: "x.com".to_string(),
            ttr: 0,
            word_count: 0,
        };
        if let Ok(xai_key) = config.require_xai_key() {
            eprintln!("Searching web for article context...");
            let http = reqwest::Client::new();
            let timeout_secs = resolve_article_timeout_secs();
            match xai::web_search_article(&http, xai_key, &url, "", &args.model, timeout_secs).await
            {
                Ok(raw) => parse_article_json(&raw, &url, "x.com", args.full),
                Err(_) => {
                    eprintln!("Could not enrich article from web. Showing metadata only.\n");
                    meta_article
                }
            }
        } else {
            eprintln!("No xAI key \u{2014} showing metadata only.\n");
            meta_article
        }
    } else {
        let xai_api_key = config.require_xai_key()?;
        let parsed = url::Url::parse(&url).map_err(|_| anyhow::anyhow!("Invalid URL: {url}"))?;
        let domain = parsed.host_str().unwrap_or("").to_string();
        let timeout_secs = resolve_article_timeout_secs();

        let http = reqwest::Client::new();
        let raw =
            xai::web_search_article(&http, xai_api_key, &url, &domain, &args.model, timeout_secs)
                .await?;
        parse_article_json(&raw, &url, &domain, args.full)
    };

    // If AI prompt provided, analyze the article
    if let Some(ai_prompt) = &args.ai {
        eprintln!("Analyzing with Grok...\n");

        let xai_key = config.require_xai_key()?;
        let http = reqwest::Client::new();
        let analysis = grok::analyze_query_tracked(
            &http,
            xai_key,
            ai_prompt,
            Some(&article.content),
            &crate::models::GrokOpts::default(),
            Some(&config.costs_path()),
        )
        .await?;

        println!("📝 Analysis: {ai_prompt}\n");
        println!("{}", analysis.content);
        println!("\n---");
    }

    if args.json {
        format::print_json_pretty_filtered(&article)?;
    } else {
        println!("{}", format_article(&article));
    }

    Ok(())
}

fn is_x_tweet_like_url(value: &str) -> bool {
    extract_tweet_id(value).is_some()
}

fn resolve_article_timeout_secs() -> u64 {
    const DEFAULT_TIMEOUT_SECS: u64 = 30;
    let parsed = std::env::var("XINT_ARTICLE_TIMEOUT_SEC")
        .ok()
        .and_then(|v| v.trim().parse::<u64>().ok())
        .unwrap_or(DEFAULT_TIMEOUT_SECS);
    parsed.clamp(5, 120)
}

fn extract_tweet_id(input: &str) -> Option<String> {
    let trimmed = input.trim();
    if trimmed.chars().all(|c| c.is_ascii_digit()) {
        return Some(trimmed.to_string());
    }

    let status_idx = trimmed.find("/status/")?;
    let id_part = &trimmed[status_idx + "/status/".len()..];
    let digits: String = id_part.chars().take_while(|c| c.is_ascii_digit()).collect();
    if digits.is_empty() {
        None
    } else {
        Some(digits)
    }
}

fn is_x_article_url(raw: &str) -> bool {
    if let Ok(url) = url::Url::parse(raw) {
        let host = url.host_str().unwrap_or_default();
        return host.eq_ignore_ascii_case("x.com") && url.path().starts_with("/i/article/");
    }
    false
}

fn is_external_non_x_url(raw: &str) -> bool {
    if let Ok(url) = url::Url::parse(raw) {
        let host = url.host_str().unwrap_or_default().to_ascii_lowercase();
        return host != "x.com"
            && host != "twitter.com"
            && !host.ends_with(".x.com")
            && !host.ends_with(".twitter.com");
    }
    false
}

fn pick_article_url_from_tweet(tweet: &Tweet) -> Option<String> {
    let mut candidates: Vec<String> = Vec::new();
    for item in &tweet.urls {
        if !item.url.is_empty() {
            candidates.push(item.url.clone());
        }
        if let Some(unwound) = &item.unwound_url {
            if !unwound.is_empty() {
                candidates.push(unwound.clone());
            }
        }
    }

    candidates.sort();
    candidates.dedup();

    if let Some(url) = candidates.iter().find(|url| is_external_non_x_url(url)) {
        return Some(url.clone());
    }
    if let Some(url) = candidates.iter().find(|url| is_x_article_url(url)) {
        return Some(url.clone());
    }
    candidates.into_iter().next()
}

async fn fetch_tweet_for_article(
    client: &XClient,
    token: &str,
    tweet_url: &str,
) -> Result<(Tweet, Option<String>, Option<Article>)> {
    let tweet_id = extract_tweet_id(tweet_url)
        .ok_or_else(|| anyhow::anyhow!("Invalid X tweet URL: {tweet_url}"))?;
    let tweet = twitter::get_tweet(client, token, &tweet_id)
        .await?
        .ok_or_else(|| anyhow::anyhow!("Tweet not found: {tweet_id}"))?;

    // If tweet has inline article data from X API, build Article directly
    if let Some(ref tw_article) = tweet.article {
        if !tw_article.plain_text.is_empty() {
            let content = reconstruct_article_content(tw_article);
            let word_count = content.split_whitespace().count() as u64;
            let ttr = (word_count as f64 / 238.0).ceil().max(1.0) as u64;
            let inline = Article {
                url: tweet_url.to_string(),
                title: if tw_article.title.is_empty() {
                    "X Article".to_string()
                } else {
                    tw_article.title.clone()
                },
                description: tw_article.preview_text.clone().unwrap_or_default(),
                content,
                author: tweet.username.clone(),
                published: tweet.created_at.clone(),
                domain: "x.com".to_string(),
                ttr,
                word_count,
            };
            return Ok((tweet, None, Some(inline)));
        }
    }

    // If tweet text is long (note_tweet was used), treat it as article content
    if tweet.text.len() > 280 {
        let word_count = tweet.text.split_whitespace().count() as u64;
        let ttr = (word_count as f64 / 238.0).ceil().max(1.0) as u64;
        let inline = Article {
            url: tweet_url.to_string(),
            title: format!("Thread by @{}", tweet.username),
            description: tweet.text.chars().take(200).collect::<String>() + "...",
            content: tweet.text.clone(),
            author: tweet.username.clone(),
            published: tweet.created_at.clone(),
            domain: "x.com".to_string(),
            ttr,
            word_count,
        };
        return Ok((tweet, None, Some(inline)));
    }

    let article_url = pick_article_url_from_tweet(&tweet);
    Ok((tweet, article_url, None))
}

fn reconstruct_article_content(article: &TweetArticle) -> String {
    let mut content = article.plain_text.clone();
    if let Some(ref entities) = article.entities {
        if let Some(ref code_blocks) = entities.code {
            if !code_blocks.is_empty() {
                content.push_str("\n\n---\n\nCode examples from article:\n");
                for block in code_blocks {
                    content.push_str(&format!("\n{}\n", block.content));
                }
            }
        }
    }
    content
}

fn parse_article_json(raw: &str, url: &str, domain: &str, full: bool) -> Article {
    // Strip markdown fences if present
    let mut cleaned = raw.trim().to_string();
    if cleaned.starts_with("```") {
        if let Some(start) = cleaned.find('\n') {
            cleaned = cleaned[start + 1..].to_string();
        }
        if cleaned.ends_with("```") {
            cleaned = cleaned[..cleaned.len() - 3].trim().to_string();
        }
    }

    let (title, description, mut content, author, published) =
        match serde_json::from_str::<serde_json::Value>(&cleaned) {
            Ok(v) => (
                v.get("title")
                    .and_then(|v| v.as_str())
                    .unwrap_or(domain)
                    .to_string(),
                v.get("description")
                    .and_then(|v| v.as_str())
                    .unwrap_or("")
                    .to_string(),
                v.get("content")
                    .and_then(|v| v.as_str())
                    .unwrap_or("")
                    .to_string(),
                v.get("author")
                    .and_then(|v| v.as_str())
                    .unwrap_or("")
                    .to_string(),
                v.get("published")
                    .and_then(|v| v.as_str())
                    .unwrap_or("")
                    .to_string(),
            ),
            Err(_) => (
                domain.to_string(),
                String::new(),
                cleaned,
                String::new(),
                String::new(),
            ),
        };

    let word_count = content.split_whitespace().count() as u64;
    let ttr = (word_count as f64 / 238.0).ceil() as u64;

    if !full && content.len() > 5000 {
        // Truncate at word boundary
        let truncated = &content[..5000];
        let end = truncated.rfind(char::is_whitespace).unwrap_or(5000);
        content = format!("{}\n\n[... truncated]", &content[..end]);
    }

    Article {
        url: url.to_string(),
        title,
        description,
        content,
        author,
        published,
        domain: domain.to_string(),
        ttr,
        word_count,
    }
}

fn format_article(article: &Article) -> String {
    let mut out = format!("\u{1f4f0} {}\n", article.title);
    if !article.author.is_empty() {
        out.push_str(&format!("\u{270d}\u{fe0f}  {}", article.author));
    }
    if !article.published.is_empty() {
        let date = if article.published.contains('T') {
            article
                .published
                .split('T')
                .next()
                .unwrap_or(&article.published)
        } else {
            &article.published
        };
        if !out.ends_with('\n') {
            out.push_str(" \u{00b7} ");
        }
        out.push_str(date);
    }
    if !article.author.is_empty() || !article.published.is_empty() {
        out.push('\n');
    }
    out.push_str(&format!("\u{1f517} {}\n", article.url));
    out.push_str(&format!(
        "\u{1f4ca} {} words \u{00b7} {} min read\n",
        article.word_count, article.ttr
    ));
    if !article.description.is_empty() {
        out.push_str(&format!("\n{}\n", article.description));
    }
    out.push_str(&format!("\n---\n\n{}", article.content));
    out
}

#[cfg(test)]
mod tests {
    use super::{extract_tweet_id, pick_article_url_from_tweet, resolve_article_timeout_secs};
    use crate::models::{Tweet, TweetMetrics, UrlEntity};

    fn fake_tweet(urls: Vec<UrlEntity>) -> Tweet {
        Tweet {
            id: "1900100012345678901".to_string(),
            text: "tweet".to_string(),
            author_id: "1".to_string(),
            username: "alice".to_string(),
            name: "Alice".to_string(),
            created_at: "2026-02-19T00:00:00Z".to_string(),
            conversation_id: "1900100012345678901".to_string(),
            metrics: TweetMetrics {
                likes: 0,
                retweets: 0,
                replies: 0,
                quotes: 0,
                impressions: 0,
                bookmarks: 0,
            },
            urls,
            mentions: vec![],
            hashtags: vec![],
            tweet_url: "https://x.com/alice/status/1900100012345678901".to_string(),
            article: None,
            organic_metrics: None,
            non_public_metrics: None,
        }
    }

    #[test]
    fn extract_tweet_id_supports_status_urls() {
        assert_eq!(
            extract_tweet_id("https://x.com/user/status/1900100012345678901"),
            Some("1900100012345678901".to_string())
        );
        assert_eq!(
            extract_tweet_id("https://twitter.com/i/web/status/1900100012345678901"),
            Some("1900100012345678901".to_string())
        );
    }

    #[test]
    fn pick_article_url_prefers_external_link() {
        let tweet = fake_tweet(vec![
            UrlEntity {
                url: "https://x.com/i/article/abc".to_string(),
                title: None,
                description: None,
                unwound_url: None,
                images: None,
            },
            UrlEntity {
                url: "https://example.com/deep-dive".to_string(),
                title: None,
                description: None,
                unwound_url: None,
                images: None,
            },
        ]);
        assert_eq!(
            pick_article_url_from_tweet(&tweet),
            Some("https://example.com/deep-dive".to_string())
        );
    }

    #[test]
    fn pick_article_url_falls_back_to_x_article() {
        let tweet = fake_tweet(vec![UrlEntity {
            url: "https://x.com/i/article/xyz".to_string(),
            title: None,
            description: None,
            unwound_url: None,
            images: None,
        }]);
        assert_eq!(
            pick_article_url_from_tweet(&tweet),
            Some("https://x.com/i/article/xyz".to_string())
        );
    }

    #[test]
    fn reconstruct_article_content_plain_text_only() {
        use crate::models::TweetArticle;
        let article = TweetArticle {
            title: "Test".to_string(),
            plain_text: "Hello world article.".to_string(),
            preview_text: None,
            cover_media: None,
            media_entities: None,
            entities: None,
        };
        assert_eq!(
            super::reconstruct_article_content(&article),
            "Hello world article."
        );
    }

    #[test]
    fn reconstruct_article_content_with_code_blocks() {
        use crate::models::{TweetArticle, TweetArticleCodeBlock, TweetArticleEntities};
        let article = TweetArticle {
            title: "Test".to_string(),
            plain_text: "Some text.".to_string(),
            preview_text: None,
            cover_media: None,
            media_entities: None,
            entities: Some(TweetArticleEntities {
                code: Some(vec![TweetArticleCodeBlock {
                    language: "rust".to_string(),
                    code: "let x = 1;".to_string(),
                    content: "```rust\nlet x = 1;\n```".to_string(),
                }]),
            }),
        };
        let result = super::reconstruct_article_content(&article);
        assert!(result.contains("Some text."));
        assert!(result.contains("Code examples from article:"));
        assert!(result.contains("```rust"));
    }

    #[test]
    fn article_timeout_defaults_and_clamps() {
        std::env::remove_var("XINT_ARTICLE_TIMEOUT_SEC");
        assert_eq!(resolve_article_timeout_secs(), 30);

        std::env::set_var("XINT_ARTICLE_TIMEOUT_SEC", "1");
        assert_eq!(resolve_article_timeout_secs(), 5);

        std::env::set_var("XINT_ARTICLE_TIMEOUT_SEC", "999");
        assert_eq!(resolve_article_timeout_secs(), 120);

        std::env::remove_var("XINT_ARTICLE_TIMEOUT_SEC");
    }
}
